docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,641 @@
+which level this pass will be performed. As const folding happens on individual
+functions, we should intuitively create a ``FunctionPass`` for it through
+``CreateFunctionPass``. The ``pass_func`` is returned as a packed function that
+invokes the ``Expr`` to ``Expr`` API on each function in a Relay module. ``{}``
+indicates that no prerequisite is required for this pass. Otherwise, the pass
+developer has to identify and list them.
+
+Meanwhile, a pass API endpoint is registered with the name
+``relay._transform.FoldConstant``. This pass, therefore, becomes an entry in the
+registry that can be accessed by both C++ (e.g. the ``GetPass`` above) and
+Python when needed.
+
+.. code:: c++
+
+    namespace transform {
+
+    Pass FoldConstant() {
+      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func =
+        [=](Function f, Module m, PassContext pc) {
+          return Downcast<Function>(FoldConstant(f));
+      };
+      return CreateFunctionPass(pass_func, 2, "FoldConstant", {});
+    }
+
+    TVM_REGISTER_API("relay._transform.FoldConstant")
+    .set_body_typed(FoldConstant);
+
+    }  // namespace transform
+
+To allow other C++ modules to apply this pass, we declare a free function in
+`include/tvm/relay/transform.h`_ as the following:
+
+.. code:: c++
+
+    TVM_DLL Pass FoldConstant();
+
+Python Frontend
+~~~~~~~~~~~~~~~
+
+Only some simple APIs are needed for the frontend side. For example, we can
+provide users the following APIs to create and execute a pass (full
+implementation is provided in `python/tvm/relay/transform.py`_). The backend
+receives the information and decides which function it should use to create
+a Pass object.
+
+PassContext
+^^^^^^^^^^^
+
+Python frontend provides a wrapper for the ``PassContext`` to enable the
+``with`` syntax by overriding ``__enter__`` and ``__exit__``. A ``current``
+static method is offered for users to get the context that is in use under
+a certain scope.
+
+.. code:: python
+
+    @register_relay_node
+    class PassContext(RelayNode):
+        def __enter__(self):
+            _transform.EnterPassContext(self)
+            return self
+    
+        def __exit__(self, ptype, value, trace):
+            _transform.ExitPassContext(self)
+    
+        @staticmethod
+        def current():
+            """Return the current pass context."""
+            return _transform.GetCurrentPassContext()
+
+A ``PassContext`` object can be instantiated through the ``build_config`` API
+which was used by Relay to configure the compilation options, including the
+optimization level, fallback device for heterogeneous execution, and
+required/disabled passes.
+
+Pass Objects
+^^^^^^^^^^^^
+
+``Pass`` is the base class of all pass objects. All methods here are just simple
+wrappers that were implemented in the backend. They are defined for users to
+conveniently interact with the base class in Python. Only a functor is defined
+to invoke the execution of a pass.
```

Comment:
This sentence is a little vague.  What functor?

Reply:
Rephrase to ``__call__``